File: src/crypto/crypto_module.py
import hashlib
import time
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import rsa, padding, ec
from cryptography.hazmat.primitives.serialization import Encoding, PrivateFormat, PublicFormat, NoEncryption
from cryptography.exceptions import InvalidSignature
from dataclasses import dataclass
from typing import List, Optional, Dict
import base64

@dataclass
class Certificate:
    subject: str
    issuer: str
    public_key: bytes
    signature: bytes
    valid_from: float
    valid_until: float

@dataclass
class SecureMessage:
    payload: bytes
    signature: bytes
    timestamp: float
    sequence_number: int
    sender_cert: Optional[bytes] = None

class CryptoModule:
    def __init__(self):
        self.private_key = None
        self.public_key = None
        self.certificate = None
        self.message_history: Dict[str, List[tuple]] = {}  # (timestamp, sequence_number, hash)
        
    def generate_key_pair(self, use_ecdsa: bool = True) -> bool:
        """Generate a new key pair using either ECDSA or RSA."""
        try:
            if use_ecdsa:
                self.private_key = ec.generate_private_key(ec.SECP256K1())
                self.public_key = self.private_key.public_key()
            else:
                self.private_key = rsa.generate_private_key(
                    public_exponent=65537,
                    key_size=2048
                )
                self.public_key = self.private_key.public_key()
            return True
        except Exception as e:
            print(f"Key generation failed: {e}")
            return False

    def hash_message(self, message: bytes, algorithm: str = 'sha256') -> bytes:
        """Hash a message using the specified algorithm."""
        if algorithm == 'sha256':
            return hashlib.sha256(message).digest()
        elif algorithm == 'md5':
            return hashlib.md5(message).digest()
        elif algorithm == 'sha1':
            return hashlib.sha1(message).digest()
        elif algorithm == 'blake2b':
            return hashlib.blake2b(message).digest()
        elif algorithm == 'sha3_256':
            return hashlib.sha3_256(message).digest()
        else:
            raise ValueError(f"Unsupported hash algorithm: {algorithm}")

    def sign_message(self, message: bytes) -> bytes:
        """Sign a message using the private key."""
        if not self.private_key:
            raise ValueError("Private key not available")

        try:
            if isinstance(self.private_key, rsa.RSAPrivateKey):
                signature = self.private_key.sign(
                    message,
                    padding.PSS(
                        mgf=padding.MGF1(hashes.SHA256()),
                        salt_length=padding.PSS.MAX_LENGTH
                    ),
                    hashes.SHA256()
                )
            else:  # ECDSA
                signature = self.private_key.sign(
                    message,
                    ec.ECDSA(hashes.SHA256())
                )
            return signature
        except Exception as e:
            print(f"Message signing failed: {e}")
            return b""

    def verify_signature(self, message: bytes, signature: bytes, public_key_bytes: bytes) -> bool:
        """Verify a message signature using a public key."""
        if not signature or not message:
            return False
            
        try:
            # Deserialize public key
            try:
                if b'BEGIN PUBLIC KEY' in public_key_bytes:
                    public_key = serialization.load_pem_public_key(public_key_bytes)
                else:
                    public_key = serialization.load_der_public_key(public_key_bytes)
            except Exception:
                # If deserialization fails, try using our own public key
                public_key = self.public_key

            if isinstance(public_key, rsa.RSAPublicKey):
                public_key.verify(
                    signature,
                    message,
                    padding.PSS(
                        mgf=padding.MGF1(hashes.SHA256()),
                        salt_length=padding.PSS.MAX_LENGTH
                    ),
                    hashes.SHA256()
                )
            else:  # ECDSA
                public_key.verify(
                    signature,
                    message,
                    ec.ECDSA(hashes.SHA256())
                )
            return True
        except (InvalidSignature, Exception) as e:
            print(f"Signature verification failed: {e}")
            return False

    def create_secure_message(self, payload: bytes) -> SecureMessage:
        """Create a secure message with signature and metadata."""
        if not payload:
            raise ValueError("Empty payload")
            
        timestamp = time.time()
        sequence_number = len(self.message_history.get(str(self.public_key), [])) + 1
        
        # Combine payload with metadata for signing
        message_data = payload + str(timestamp).encode() + str(sequence_number).encode()
        signature = self.sign_message(message_data)
        
        if not signature:
            raise ValueError("Failed to sign message")
        
        # Create secure message
        secure_msg = SecureMessage(
            payload=payload,
            signature=signature,
            timestamp=timestamp,
            sequence_number=sequence_number
        )
        
        # Add certificate if available
        if self.certificate:
            secure_msg.sender_cert = self.certificate
            
        # Update message history
        if str(self.public_key) not in self.message_history:
            self.message_history[str(self.public_key)] = []
        self.message_history[str(self.public_key)].append(
            (timestamp, sequence_number, self.hash_message(payload))
        )
        
        return secure_msg

    def verify_secure_message(self, message: SecureMessage) -> bool:
        """Verify a secure message's integrity and authenticity."""
        try:
            # Check timestamp (5 second tolerance)
            if abs(time.time() - message.timestamp) > 5:
                return False
                
            # Check for replay
            if self.is_replay_message(message):
                return False
                
            # Verify certificate if present
            if message.sender_cert:
                cert = self._deserialize_certificate(message.sender_cert)
                if not self._verify_certificate(cert):
                    return False
            
            # Combine message data for verification
            message_data = message.payload + str(message.timestamp).encode() + str(message.sequence_number).encode()
            
            # Get public key for verification
            public_key_bytes = message.sender_cert if message.sender_cert else (
                self.public_key.public_bytes(
                    encoding=Encoding.DER,
                    format=PublicFormat.SubjectPublicKeyInfo
                ) if self.public_key else None
            )
            
            if not public_key_bytes:
                return False
                
            return self.verify_signature(message_data, message.signature, public_key_bytes)
        except Exception as e:
            print(f"Message verification failed: {e}")
            return False

    def is_replay_message(self, message: SecureMessage) -> bool:
        """Check if a message is a replay attack."""
        for timestamp, seq, msg_hash in self.message_history.get(str(self.public_key), []):
            if (message.timestamp == timestamp and 
                message.sequence_number == seq and 
                self.hash_message(message.payload) == msg_hash):
                return True
        return False

    def _deserialize_certificate(self, cert_bytes: bytes) -> Certificate:
        """Deserialize a certificate from bytes."""
        # Implementation depends on your certificate format
        # This is a simplified example
        return Certificate(
            subject="test",
            issuer="test",
            public_key=cert_bytes,
            signature=b"",
            valid_from=time.time(),
            valid_until=time.time() + 3600
        )

    def _verify_certificate(self, cert: Certificate) -> bool:
        """Verify a certificate's validity."""
        current_time = time.time()
        return cert.valid_from <= current_time <= cert.valid_until 

File: src/crypto/crypto-module.cpp
#include "crypto-module.h"
#include <chrono>
#include <algorithm>
#include <stdexcept>
#include <cstring>

namespace vanet {
namespace crypto {

// Constants for security parameters
constexpr size_t MAX_MESSAGE_HISTORY = 1000;  // Maximum number of messages to store for replay prevention
constexpr uint64_t MESSAGE_TIMEOUT = 5000;    // Message timeout in milliseconds
constexpr size_t MIN_KEY_SIZE = 2048;         // Minimum RSA key size in bits
constexpr size_t MAX_CERT_CHAIN = 5;          // Maximum depth of certificate chain

CryptoModule::CryptoModule() : privateKey(nullptr), publicKey(nullptr), certificate(nullptr) {
    initializeOpenSSL();
}

CryptoModule::~CryptoModule() {
    cleanupOpenSSL();
}

void CryptoModule::initializeOpenSSL() {
    OpenSSL_add_all_algorithms();
    ERR_load_crypto_strings();
}

void CryptoModule::cleanupOpenSSL() {
    if (privateKey) EVP_PKEY_free(privateKey);
    if (publicKey) EVP_PKEY_free(publicKey);
    if (certificate) X509_free(certificate);
    EVP_cleanup();
    ERR_free_strings();
}

bool CryptoModule::generateKeyPair(SignatureAlgorithm algo) {
    EVP_PKEY_CTX* ctx = nullptr;
    
    if (algo == SignatureAlgorithm::RSA_PSS) {
        ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_RSA, nullptr);
        if (!ctx) return false;
        
        if (EVP_PKEY_keygen_init(ctx) <= 0) {
            EVP_PKEY_CTX_free(ctx);
            return false;
        }
        
        if (EVP_PKEY_CTX_set_rsa_keygen_bits(ctx, MIN_KEY_SIZE) <= 0) {
            EVP_PKEY_CTX_free(ctx);
            return false;
        }
    } else { // ECDSA
        ctx = EVP_PKEY_CTX_new_id(EVP_PKEY_EC, nullptr);
        if (!ctx) return false;
        
        if (EVP_PKEY_keygen_init(ctx) <= 0) {
            EVP_PKEY_CTX_free(ctx);
            return false;
        }
        
        if (EVP_PKEY_CTX_set_ec_paramgen_curve_nid(ctx, NID_secp256k1) <= 0) {
            EVP_PKEY_CTX_free(ctx);
            return false;
        }
    }
    
    EVP_PKEY* key = nullptr;
    if (EVP_PKEY_keygen(ctx, &key) <= 0) {
        EVP_PKEY_CTX_free(ctx);
        return false;
    }
    
    if (privateKey) EVP_PKEY_free(privateKey);
    privateKey = key;
    EVP_PKEY_CTX_free(ctx);
    return true;
}

std::vector<uint8_t> CryptoModule::hashMessage(const std::vector<uint8_t>& message, HashAlgorithm algo) {
    const EVP_MD* md;
    switch (algo) {
        case HashAlgorithm::SHA256:
            md = EVP_sha256();
            break;
        case HashAlgorithm::MD5:
            md = EVP_md5();
            break;
        case HashAlgorithm::SHA1:
            md = EVP_sha1();
            break;
        case HashAlgorithm::BLAKE2B:
            md = EVP_blake2b512();
            break;
        case HashAlgorithm::SHA3_256:
            md = EVP_sha3_256();
            break;
        default:
            throw std::invalid_argument("Unsupported hash algorithm");
    }

    EVP_MD_CTX* mdctx = EVP_MD_CTX_new();
    std::vector<uint8_t> hash(EVP_MAX_MD_SIZE);
    unsigned int hashLen;

    EVP_DigestInit_ex(mdctx, md, nullptr);
    EVP_DigestUpdate(mdctx, message.data(), message.size());
    EVP_DigestFinal_ex(mdctx, hash.data(), &hashLen);
    EVP_MD_CTX_free(mdctx);

    hash.resize(hashLen);
    return hash;
}

std::vector<uint8_t> CryptoModule::signMessage(const std::vector<uint8_t>& message) {
    if (!privateKey) {
        throw std::runtime_error("Private key not loaded");
    }

    EVP_MD_CTX* ctx = EVP_MD_CTX_new();
    if (!ctx) {
        throw std::runtime_error("Failed to create signature context");
    }

    if (EVP_DigestSignInit(ctx, nullptr, EVP_sha256(), nullptr, privateKey) <= 0) {
        EVP_MD_CTX_free(ctx);
        throw std::runtime_error("Failed to initialize signature");
    }

    size_t sigLen;
    if (EVP_DigestSignUpdate(ctx, message.data(), message.size()) <= 0 ||
        EVP_DigestSignFinal(ctx, nullptr, &sigLen) <= 0) {
        EVP_MD_CTX_free(ctx);
        throw std::runtime_error("Failed to calculate signature length");
    }

    std::vector<uint8_t> signature(sigLen);
    if (EVP_DigestSignFinal(ctx, signature.data(), &sigLen) <= 0) {
        EVP_MD_CTX_free(ctx);
        throw std::runtime_error("Failed to create signature");
    }

    EVP_MD_CTX_free(ctx);
    signature.resize(sigLen);
    return signature;
}

bool CryptoModule::verifySignature(const std::vector<uint8_t>& message,
                                 const std::vector<uint8_t>& signature,
                                 const std::vector<uint8_t>& publicKey) {
    EVP_PKEY* key = nullptr;
    const unsigned char* pubKeyData = publicKey.data();
    key = d2i_PUBKEY(nullptr, &pubKeyData, publicKey.size());
    
    if (!key) {
        return false;
    }

    EVP_MD_CTX* ctx = EVP_MD_CTX_new();
    bool result = false;

    if (EVP_DigestVerifyInit(ctx, nullptr, EVP_sha256(), nullptr, key) > 0 &&
        EVP_DigestVerifyUpdate(ctx, message.data(), message.size()) > 0 &&
        EVP_DigestVerifyFinal(ctx, signature.data(), signature.size()) > 0) {
        result = true;
    }

    EVP_PKEY_free(key);
    EVP_MD_CTX_free(ctx);
    return result;
}

CryptoModule::SecureMessage CryptoModule::createSecureMessage(const std::vector<uint8_t>& payload) {
    SecureMessage msg;
    msg.payload = payload;
    msg.timestamp = std::chrono::duration_cast<std::chrono::milliseconds>(
        std::chrono::system_clock::now().time_since_epoch()
    ).count();
    
    static uint32_t sequence = 0;
    msg.sequenceNumber = ++sequence;
    
    // Create signature over payload + timestamp + sequence number
    std::vector<uint8_t> toSign = payload;
    toSign.insert(toSign.end(), 
                 reinterpret_cast<uint8_t*>(&msg.timestamp),
                 reinterpret_cast<uint8_t*>(&msg.timestamp) + sizeof(msg.timestamp));
    toSign.insert(toSign.end(),
                 reinterpret_cast<uint8_t*>(&msg.sequenceNumber),
                 reinterpret_cast<uint8_t*>(&msg.sequenceNumber) + sizeof(msg.sequenceNumber));
    
    msg.signature = signMessage(toSign);
    
    // Add certificate if available
    if (certificate) {
        unsigned char* certBuf = nullptr;
        int certLen = i2d_X509(certificate, &certBuf);
        if (certLen > 0) {
            msg.senderCert.assign(certBuf, certBuf + certLen);
            OPENSSL_free(certBuf);
        }
    }
    
    return msg;
}

bool CryptoModule::verifySecureMessage(const SecureMessage& message) {
    // Check timestamp
    if (!isValidTimestamp(message.timestamp)) {
        return false;
    }
    
    // Check for replay
    if (isReplayMessage(message)) {
        return false;
    }
    
    // Verify certificate if present
    if (!message.senderCert.empty()) {
        const unsigned char* certData = message.senderCert.data();
        X509* cert = d2i_X509(nullptr, &certData, message.senderCert.size());
        if (!cert || !verifyCertificate(Certificate{/*...*/})) {
            X509_free(cert);
            return false;
        }
        X509_free(cert);
    }
    
    // Verify signature
    std::vector<uint8_t> toVerify = message.payload;
    toVerify.insert(toVerify.end(),
                   reinterpret_cast<const uint8_t*>(&message.timestamp),
                   reinterpret_cast<const uint8_t*>(&message.timestamp) + sizeof(message.timestamp));
    toVerify.insert(toVerify.end(),
                   reinterpret_cast<const uint8_t*>(&message.sequenceNumber),
                   reinterpret_cast<const uint8_t*>(&message.sequenceNumber) + sizeof(message.sequenceNumber));
    
    return verifySignature(toVerify, message.signature, message.senderCert);
}

bool CryptoModule::isReplayMessage(const SecureMessage& message) {
    auto it = std::find_if(messageHistory.begin(), messageHistory.end(),
        [&message](const MessageHistory& history) {
            return history.timestamp == message.timestamp &&
                   history.sequenceNumber == message.sequenceNumber;
        });
    return it != messageHistory.end();
}

void CryptoModule::updateMessageHistory(const SecureMessage& message) {
    MessageHistory history;
    history.timestamp = message.timestamp;
    history.sequenceNumber = message.sequenceNumber;
    history.messageHash = hashMessage(message.payload);
    
    messageHistory.push_back(history);
    
    // Prune old messages if necessary
    if (messageHistory.size() > MAX_MESSAGE_HISTORY) {
        pruneMessageHistory();
    }
}

void CryptoModule::pruneMessageHistory() {
    auto now = std::chrono::duration_cast<std::chrono::milliseconds>(
        std::chrono::system_clock::now().time_since_epoch()
    ).count();
    
    messageHistory.erase(
        std::remove_if(messageHistory.begin(), messageHistory.end(),
            [now](const MessageHistory& history) {
                return (now - history.timestamp) > MESSAGE_TIMEOUT;
            }),
        messageHistory.end()
    );
}

bool CryptoModule::isValidTimestamp(uint64_t timestamp) const {
    auto now = std::chrono::duration_cast<std::chrono::milliseconds>(
        std::chrono::system_clock::now().time_since_epoch()
    ).count();
    
    return (now - timestamp) <= MESSAGE_TIMEOUT;
}

} // namespace crypto
} // namespace vanet 